clang/lib/Frontend/CodeSnippetHighlighter.cpp

```diff
@@ -0,0 +1,185 @@
+//===-- CodeSnippetHighlighter.cpp - Code snippet highlighting --*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "clang/Frontend/CodeSnippetHighlighter.h"
+#include "clang/Basic/CharInfo.h"
+#include "clang/Basic/DiagnosticOptions.h"
+#include "clang/Basic/SourceManager.h"
+#include "clang/Lex/Lexer.h"
+#include "clang/Lex/Preprocessor.h"
+#include "clang/Lex/PreprocessorOptions.h"
+#include "llvm/Support/raw_ostream.h"
+#include <chrono>
+
+using namespace clang;
+
+// Magenta is taken for 'warning'. Red is already 'error' and 'cyan'
+// is already taken for 'note'. Green is already used to underline
+// source ranges. White and black are bad because of the usual
+// terminal backgrounds. Which leaves us only with TWO options.
+static constexpr raw_ostream::Colors CommentColor = raw_ostream::YELLOW;
+static constexpr raw_ostream::Colors LiteralColor = raw_ostream::GREEN;
+static constexpr raw_ostream::Colors KeywordColor = raw_ostream::BLUE;
+/// Maximum size of file we still highlight.
+static constexpr size_t MaxBufferSize = 1024 * 1024; // 1MB.
+
+llvm::SmallVector<StyleRange> CodeSnippetHighlighter::highlightLine(
+    unsigned LineNumber, const Preprocessor *PP, const LangOptions &LangOpts,
+    FileID FID, const SourceManager &SM, const char *LineStart) {
+  std::chrono::steady_clock::time_point begin =
+      std::chrono::steady_clock::now();
+
+  if (!PP)
+    return {};
+
```

Comment:
Under what circumstances is `PP` null? Can we turn this into an assert? Or better yet, can we turn it into a const reference?

Reply:
`Preprocessor*` is what's passed to `TextDiagnosticPrinter::BeginSourceFile()`, and there are a couple of places that pass `nullptr`, here an excerpt from `ag`:
```
../clang/lib/Frontend/FrontendAction.cpp
580:bool FrontendAction::BeginSourceFile(CompilerInstance &CI,
699:    CI.getDiagnosticClient().BeginSourceFile(CI.getLangOpts(), nullptr);
715:    if (!BeginSourceFileAction(CI))
759:    CI.getDiagnosticClient().BeginSourceFile(CI.getLangOpts(), nullptr);
763:    if (!BeginSourceFileAction(CI))
815:  CI.getDiagnosticClient().BeginSourceFile(CI.getLangOpts(),
914:  if (!BeginSourceFileAction(CI))
1207:bool WrapperFrontendAction::BeginSourceFileAction(CompilerInstance &CI) {
1210:  auto Ret = WrappedAction->BeginSourceFileAction(CI);
1211:  // BeginSourceFileAction may change CurrentInput, e.g. during module builds.

``

Reply:
Thanks! I guess this stays then. :-)